Add Minimum and Maximum blend modes to CanvasItem and Spatial shaders

Calinou opened the pull request:
`master` version of https://github.com/godotengine/godot/pull/48004.

Differences from the `3.x` PR:

- The final result looks different (see https://github.com/godotengine/godot/pull/48004#issuecomment-839146688). Which one is correct?
  - **Edit:** It seems this PR's result is the correct one.
- ~~The blend mode enum is ordered differently to ensure the "premultiplied alpha" and "disabled" options are last. This breaks compatibility with existing materials, but is more consistent.~~
  - **Edit:** This is no longer done to preserve backwards compatibility.
- The enum value and blend mode is named "Minimum" and "Maximum" instead of "Min" and "Max". This avoids a conflict with `_MAX` being used within Godot to denote the highest possible value of an enum.

These blend modes are useful for some special effects.

This closes https://github.com/godotengine/godot-proposals/issues/2621.

**Testing project:** [test_minimum_maximum_blend_modes.zip](https://github.com/user-attachments/files/24117617/test_minimum_maximum_blend_modes.zip)

## Preview

*From left to right: Minimum, Maximum*

### 2D

Forward+ | Mobile | Compatibility
-|-|-
<img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/7d133780-82e2-408d-9f2b-1b4244844023" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/07005529-3dd5-4351-984a-0a72092f0aeb" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/baae7d49-725e-4e0f-8042-469279a9f127" />

### 3D

Forward+ | Mobile | Compatibility
-|-|-
<img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/7f0030b6-3e5b-4e38-af88-9f3e1ee1a251" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/92ee2760-48a2-4eb6-8d7d-8f24b26801b9" /> | <img width="1152" height="648" alt="Image" src="https://github.com/user-attachments/assets/ec77e3da-7eea-40a5-bfe4-c515c7b3f026" />

Calinou committed:
Add Minimum and Maximum blend modes to CanvasItem and Spatial shaders

These blend modes are useful for some special effects.

reduz commented:
This would be good to merge. It was lacking because of GLES2 compatilbity but it should be ok to do now.
It would be need to be salvaged.

QbieShay commented:
Is this possible to be included in 4.x without breaking compat?

Calinou commented:
Rebased and tested again, it works as expected. I've changed the enum order to preserve backwards compatibility.

This works in all rendering methods in 2D and 3D, but note that in OpenGL, 3D materials must explicitly use alpha transparency for the Minimum and Maximum blend modes to render correctly. I'm not sure how I can declare these blend modes to implicitly require alpha transparency as is already done for the Add, Subtract and Multiply blend modes.

clayjohn commented:
For GLES3, you need to set the ``uses_blend_alpha`` variable 

https://github.com/godotengine/godot/blob/eb86dabee07e8dfce3b06cbd557b50b74afd3d6c/drivers/gles3/storage/material_storage.cpp#L3016-L3018

kleonc commented:
Needs rebasing.

Based on https://github.com/godotengine/godot/pull/48654#issuecomment-1199747556 this is already accepted to be added in, is there anything else to be done here besides reviewing/testing?
Would be nice to get this finally merged in (and makes sense doing so early in the dev cycle).

QbieShay commented:
I'm happy to see this merged, however it seems that the blend modes for spatial are added before the alpha to coverage.

Since we're working to add premul alpha, would be nice if this PR can also introduce the premul alpha constant before the min/max, so we can keep consistency with canvas.

Calinou commented:
> Since we're working to add premul alpha, would be nice if this PR can also introduce the premul alpha constant before the min/max, so we can keep consistency with canvas.

https://github.com/godotengine/godot/pull/68548 would need to be merged first, so this can be done without breaking compatibility (even if we're during the dev cycle where this isn't critical, it's preferred).

QbieShay commented:
Alright. Im currently blocked on a couple of design decision we'll probably makein wed's meeting, then i can move forward with implementation

xne commented:
Now that #85609 has been merged, this just needs testing and review?

QbieShay commented:
This PR definitely needs a rebase before it can be tested - but we're in feature freeze right now, so this is 4.5 material anyway.

Calinou commented:
> Is this still planned for 4.5?

4.5 is now in feature freeze, so new features can only be merged in 4.6 at the earliest.

Calinou commented:
Rebased and tested again, it works as expected on all renderers. Thanks @Macksaur for the pointers :slightly_smiling_face: 

Unfortunately, this missed the 4.6 merge window (as it's [now in feature freeze](https://godotengine.org/article/dev-snapshot-godot-4-6-dev-6/)), so this will have to wait for a future 4.x release. If you need this feature right now, you can [compile](https://docs.godotengine.org/en/latest/engine_details/development/compiling/index.html) custom editor and export template binaries with this PR included.
